### PR TITLE
Do minimal invalidation with soft breaks enabled

### PR DESCRIPTION
**Pull request by cmyr**
This is a little complicated, because we need different information
after an edit than we do when we're just rewrapping. Most importantly,
when an edit occurs it's possible that the edit deletes existing
breaks; this means we need to count the breaks in the pre-edit
rope and breaks structures in order to handle invalidation correctly.

----

progress on #1053

## Review Checklist

- [ ] I have responded to reviews and made changes where appropriate.
- [x] I have tested the code with `cargo test --all` / `./rust/run_all_checks`.
- [x] I have updated comments / documentation related to the changes I made.
- [x] I have rebased my PR branch onto xi-editor/master.

### CODE REVIEW

File: rust/core-lib/src/linewrap.rs [hunk 1]
```diff
@@ -82,6 +82,21 @@ pub(crate) struct InvalLines {
...
+/// the information relevant to both of them.
+struct WrapSummary {
+    start_line: usize,
+    /// Total number of invalidated lines; this is meaningless in the after_edit case.
+    inval_count: usize,
+    /// The total number of new (hard + soft) breaks in the wrapped region.
+    new_count: usize,
+    /// The number of new soft breaks.
+    new_soft: usize,
+}
```
Comment (jansol) [inline rust/core-lib/src/linewrap.rs]: This looks like a C-ism. Is any of the other fields unique to one case? If so, I'd probably turn this into a struct enum just to err on the safe side. If not, maybe consider making `invalid_count` a `Option<usize>`?

Comment (cmyr) [reply rust/core-lib/src/linewrap.rs]: The problem is  the function we're calling doesn't know which of the two callers it's dealing with. I've taken the approach of just making sure I point out the tricky bits and hopefully explain the situation. Doesn't feel great or anything but felt like the best option. 😞

Comment (jansol) [reply rust/core-lib/src/linewrap.rs]: Hmm yeah, reading the code it makes a bit more sense to have it be this way.

**Comment by jansol** [comment]
Feels like nice speedup in sqlite3.c already.

**Pull request merged**